cloudmarker/alerts/emailalert.py

```diff
@@ -68,10 +79,53 @@ def done(self):
             self.body = ''.join(self.stringbuffer)
         message.attach(MIMEText(self.body))
 
-        smtp_connection = smtplib.SMTP(host=self.host, port=self.port)
+        smtp_connection = self.get_connection()
         try:
             smtp_connection.sendmail(self.sender, self.to, message.as_string())
         except smtplib.SMTPException as e:
             _log.error('Failed to send email: %s', e)
         finally:
             smtp_connection.quit()
+
+    def get_connection(self):
+        """Return SMTP connection object.
+
+        Create a SMTP connection based on whether the SSL param.
+        If SSL connection is required (for eg: in case of gmail) then
+        method will return a SMTP_SSL connection. In other cases if the
+        authentication is not required or SMTP_SSL connection is not
+        required then a plain SMTP connection object is returned.
+        # For gmail follow the steps given below even to work for 2 factor
+        authentication
+        # 1. Log-in into Gmail with your account
+        # 2. Navigate to https://security.google.com/settings/security/
+        apppasswords
+        # 3. In 'select app' choose 'custom', give it an arbitrary name and
+        press generate
+        # 4. It will give you 16 chars token.
+        # 5. Use that token as password for login
+        # 6. Host is smtp.gmail.com and port is 465
+        """
+        # In case no authentication is required then no need to login
+        if self.is_ssl_required:
+
+            smtp_connection = smtplib.SMTP_SSL(host=self.host, port=self.port)
+            smtp_connection.set_debuglevel(True)
+
+            # identity yourself to esmtp server using ehlo. This call is
+            # required
+            # or else it would throw error: SMTPSenderRefused: (503, b'5.5.1
+            # EHLO/HELO first.
+            smtp_connection.ehlo()
+            smtp_connection.connect(self.host, self.port)
+
+            # an ehlo call is required just after the connect method and just
+            # before login
+            # just to identity yourself. If not called it would throw a
+            # SMTPSenderRefused error
+            # Failed to send email; error: SMTPSenderRefused: (503, b'5.5.1
+            # EHLO/HELO first.
+            smtp_connection.ehlo()
+            smtp_connection.login(self.username, self.password)
+            return smtp_connection
+        return smtplib.SMTP(host=self.host, port=self.port)
```

Comment:
If the code is simplified as described in the previous comments, then this whole separate function would no longer be required. It would be a simple 4-line if-then code which can just be moved back to the `write()` method.

Reply:
I have removed the `ehlo` methods but I am keeping the `_get_connection` method. The reason for doing so is that it creates the connection based on the `use_ssl` params and returns the object. Also it contains the comments to do so. 

So I believe its better to keep this as a separate method rather then flooding the `write` method.

Reply:
This sounds fair.

I think the function name should be updated. The `_get_connection()` method is not merely getting a connection. It is doing much more than that. It sends the `EHLO` payload behind the scenes. It also performs authentication with `smtp.login()`. This is more of a helper method than something that is merely establishing a connection. We should update the method name to accurately reflect this. Some ideas could be `_start_smtp_session()`, `_prepare_smtp_session()`, `_smtp_session()`, etc.

Reply:
The `_get_connection()` means a function which returns a smtp connection object. I will rename the function `_prepare_smtp_session()` since it looks more relative to the functionality.

Reply:
Sounds good.

I am curious if you are also planning to work on #53 (Send email notification after all audits have run) later. If so, this `_prepare_smtp_session()` method has to be moved out to another module outside the plugin packages. The reason is that the change for #53 would be in the framework code (`manager.py` or `workers.py`). The framework cannot depend on individual plugins for its own functionality. So the framework would also have to repeat this work of creating an SMTP connection and starting an SMTP session before it can send email notification related to audits (e.g., an email notification that says that the audits have completed).

So instead of repeating the `_prepare_smtp_session()` code twice (once in `EmailAlert` and once more in the framework code in `manager.py`/`worker.py`), it would make sense to have this `_prepare_smtp_session()` functionality moved to a common place such as `util.py` that both framework and `EmailAlert` can reuse.

Of course, this does not have to be done right now in this pull request. This pull request may remain self contained and not worry about issue #53. What I said above has to be done sooner or later when you or someone else works on #53.